### DISCUSSION TITLE
Similarity Search Returning Null Result (Milvus, NodeJS) (PLEASE HELP)

### DISCUSSION THREAD

**Discussion by razataiab** [milvus-io/milvus#33809] [Q&A and General discussion]
I am trying to Implement a RAG using Milvus, Langchain and NodeJS. The Embedding for the Text and Query are being generated but upon performing a similarity search using both of their embeddings, nothings being returned.

PS: The main code for searching is in similaritySearch.js, the code for creating a collection and inserting embedding is in milvus.js, the code for generating the embeddings is in processing.js

milvus.js:

const { MilvusClient, DataType } = require('@zilliz/milvus2-sdk-node');
const express = require('express');

const app = express();
app.use(express.json());

const client = new MilvusClient("localhost:19530");

function generateUniqueCollectionName() {
    const timestamp = Date.now();
    return `embedding_${timestamp}`;
}

async function storeEmbeddings(collectionName, filename, embeddings) {
    try {
        console.log(`Embedding length: ${embeddings.length}`);

        await client.createCollection({
            collection_name: collectionName,
            fields: [
                { name: "my_id", data_type: DataType.Int64, is_primary_key: true, auto_id: true },
                { name: "filename", data_type: DataType.VarChar, max_length: 256 },
                { name: "embedding", data_type: DataType.FloatVector, dim: embeddings[0].length }
            ],
        });
        console.log(`Collection ${collectionName} created`);

        // Prepare the row-based data for insertion
        const rowData = embeddings.map(embedding => ({
            filename: filename,
            embedding: embedding
        }));

        await client.insert({
            collection_name: collectionName,
            data: rowData
        });

        await client.createIndex({
            collection_name: collectionName,
            field_name: "embedding",
            index_type: "IVF_FLAT",
            metric_type: "L2",
            index_name: "vector_index",
            params: { "nlist": 128 }
        })

        console.log("Embeddings stored successfully in Milvus");
    } catch (error) {
        console.error('Failed to store embeddings in Milvus:', error);
    }
}

module.exports = {
    storeEmbeddings,
    generateUniqueCollectionName,
};

query.js:

const express = require('express');
const { generateEmbedding } = require('../utils/processing');
const { similaritySearch } = require('../utils/similaritySearch.js');
const router = express.Router();

router.post('/', async (req, res) => {
    try {
        const { query } = req.body;

        if (!query) {
            return res.status(400).json({ error: 'Query text is required' });
        }

        // Generate embedding for the query text
        const embedding = await generateEmbedding(query);

        console.log('Embedding for query generated successfully');
        console.log('Embedding dimension:', embedding.length);

        // Perform similarity search
        const results = await similaritySearch(embedding);

        res.json({ results });
    } catch (error) {
        console.error('Error generating embedding for query:', error);
        res.status(500).json({ error: 'Internal server error' });
    }
});

module.exports = router;

similaritySearch.js:

const { MilvusClient } = require('@zilliz/milvus2-sdk-node');
const { getCollectionName } = require('../utils/config'); // Import the config module

const client = new MilvusClient({ address: 'localhost:19530' });

async function similaritySearch(queryEmbedding) {
    try {
        const collectionName = getCollectionName(); // Retrieve the collection name
        if (!collectionName) {
            throw new Error('Collection name is not set.');
        }

        await client.loadCollection({
            collection_name: collectionName
        });

        // Conduct similarity search
        const res = await client.search({
            collection_name: collectionName,
            vectors: [queryEmbedding],
            topk: 5,
        });

        // Check if response is defined and has results
        if (!res || !res.results) {
            throw new Error('Search response is undefined or does not contain results');
        }

        console.log('Similarity search response:', res.results);
        return res.results;
    } catch (error) {
        console.error('Error during similarity search:', error);
        throw error;
    }
}

module.exports = { similaritySearch };

processing.js:

const fs = require('fs').promises;
const { OpenAI } = require('openai');
const { RecursiveCharacterTextSplitter } = require('langchain/text_splitter');

const openai = new OpenAI({ apiKey: 'sk-proj-o70B1pGGGDU5fg1as1DcT3BlbkFJOiMIcI2ocj15oSJO5gLQ' });

async function processDocument(filePath) {
    try {
        const text = await fs.readFile(filePath, 'utf-8');
        console.log(`Text from file: ${text}`);
        console.log(`Text length: ${text.length}`);

        const chunks = await splitDocumentIntoChunks(text);
        
        const embeddings = await generateEmbeddingsForChunks(chunks);
        return embeddings;
    } catch (error) {
        console.error('Error processing document:', error);
        throw error;
    }
}

async function splitDocumentIntoChunks(text) {
    const textSplitter = new RecursiveCharacterTextSplitter({ chunkSize: 50, chunkOverlap: 10 });

    const docOutput = await textSplitter.createDocuments([text]);

    // Extract page content from each document chunk
    const chunkTexts = docOutput.map(doc => doc.pageContent);
    console.log('Chunks:', chunkTexts);

    return chunkTexts;
}

async function generateEmbeddingsForChunks(chunks) {
    const embeddings = [];

    for (const chunk of chunks) {
        const embedding = await generateEmbedding(chunk);
        embeddings.push(embedding);
    }

    return embeddings;
}

async function generateEmbedding(text) {
    const response = await openai.embeddings.create({
        model: "text-embedding-3-small",
        input: [text],
        encoding_format: "float",
    });

    console.log('Embedding generated:', response.data[0].embedding);

    return response.data[0].embedding;
}

module.exports = {
    processDocument,
    generateEmbedding
};

**Comment by yhmo** [comment]
Set the consistency_level to be Strong and try again.
```
const res = await client.search({
        collection_name: collectionName,
        vectors: [queryEmbedding],
        topk: 5,
        consistency_level: ConsistencyLevelEnum.Strong,
    });
```

Consistency level controls the data visibility for search/query.
https://milvus.io/docs/consistency.md#Consistency

**Reply by razataiab** [reply]
still a null result

**Comment by shanghaikid** [comment]
can you try attu to check if your data has been inserted? 
https://github.com/zilliztech/attu

**Reply by razataiab** [reply]
I've tried to access the client but it's just not working on the web even though milvus and Attu are on the same docker network

**Reply by shanghaikid** [reply]
ok, you are using attu docker version.  

Make sure the milvus address you enter on the web can be accessed from the attu docker container.   

if you enter `127.0.0.1:19530`, the attu container will try to connect to `127.0.0.1:19530`, it should not work, because it is the local networking address in the attu container.

**Reply by shanghaikid** [reply]
if you have trouble to understand the networking issue, you can use desktop version.  download it from there https://github.com/zilliztech/attu/releases/tag/v2.4.1

**Reply by yhmo** [reply]
Export the port of milvus container in the docker-compose.yaml:
```
  standalone:
    container_name: milvus-standalone
    ......
    ports:
      - "19530:19530"
      - "9091:9091"
```

Use the local IP to connect milvus in the Attu login page:
![Screenshot from 2024-06-13 16-50-59](https://github.com/milvus-io/milvus/assets/2282099/fd165a3c-afff-4fc3-881f-5fdb2494cfd2)

**Reply by razataiab** [reply]
<img width="1440" alt="Screenshot 2024-06-13 at 10 02 06" src="https://github.com/milvus-io/milvus/assets/124519821/4ac4a515-bccb-4caf-b4e2-23ceb4634ebb">


This was my command at terminal: docker run -p 8000:3000 -e MILVUS_URL={10.193.148.138}:19530 zilliz/attu:v2.4.0

**Reply by shanghaikid** [reply]
@yhmo `milvus is not ready yet` means `CheckHealth` failed.

**Reply by razataiab** [reply]
so how to fix it?

**Reply by shanghaikid** [reply]
how do you start your milvus? docker?

**Reply by razataiab** [reply]
I just open docker desktop and run the containers

**Reply by razataiab** [reply]
<img width="1440" alt="Screenshot 2024-06-13 at 10 15 49" src="https://github.com/milvus-io/milvus/assets/124519821/81bcd8c7-9437-4c51-90a7-21dbf35f2b23">

**Reply by shanghaikid** [reply]
> <img alt="Screenshot 2024-06-13 at 10 15 49" width="1440" src="https://private-user-images.githubusercontent.com/124519821/339299332-81bcd8c7-9437-4c51-90a7-21dbf35f2b23.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyNzA1OTYsIm5iZiI6MTcxODI3MDI5NiwicGF0aCI6Ii8xMjQ1MTk4MjEvMzM5Mjk5MzMyLTgxYmNkOGM3LTk0MzctNGM1MS05MGE3LTIxZGJmMzVmMmIyMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxM1QwOTE4MTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MzVjYmQxMjEzMDE5ZTY4YjAwMDlhZGJiMWYyYTFiOWUzNzljNTgyOTAyMzM3MDkzMjc4NmE0NGY2MmY4OTMxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.irdX26h_fW3ZFluFfCRBPC-z9oqVARyycVyCcKaUcOY">

This means attu can not connect to the  milvus server.

**Comment by shanghaikid** [comment]
anyway, can you print this

```
const insert =   await client.insert({
        collection_name: collectionName,
        data: rowData
    });

console.log('insert result`, insert)
```

**Reply by shanghaikid** [reply]
I guess, you should use 
```
 await client.loadCollectionSync({
        collection_name: collectionName
    });
```

**Reply by razataiab** [reply]
razataiab@Razas-MacBook-Pro rag-web-app % node backend/server.js
(node:42135) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
Server running on http://localhost:3001
Processing file: f94576f715fa7cd7135b00ab0e0cf032
Text from file: The Co-operative brand is also bigger than just food and grocery delivery, you also provide services for insurance, funeral care, and legal advice. A key difference from other brands, such as other retail stores, is that a co-operative (co-op) is a unique type of company. I am aware that individual members and other co-ops own your cooperative, not large investors, and your members have a voice in how you operate.


Text length: 419
Chunks: [
  'The Co-operative brand is also bigger than just',
  'than just food and grocery delivery, you also',
  'you also provide services for insurance, funeral',
  'funeral care, and legal advice. A key difference',
  'from other brands, such as other retail stores,',
  'stores, is that a co-operative (co-op) is a',
  'is a unique type of company. I am aware that',
  'that individual members and other co-ops own your',
  'own your cooperative, not large investors, and',
  'and your members have a voice in how you operate.'
]


Embeddings generated for file: f94576f715fa7cd7135b00ab0e0cf032
Embedding length: 1536
Collection embedding_1718271366449 created
insert result:  {
  succ_index: [],
  err_index: [
    0, 1, 2, 3, 4,
    5, 6, 7, 8, 9
  ],
  status: {
    extra_info: {},
    error_code: 'UnexpectedError',
    reason: 'the num_rows(0) of 0th field is not equal to passed NumRows(10)',
    code: 0,
    retriable: false,
    detail: ''
  },
  IDs: null,
  acknowledged: false,
  insert_cnt: '0',
  delete_cnt: '0',
  upsert_cnt: '0',
  timestamp: '0'
}
Embeddings stored successfully in Milvus
Embeddings stored for file: f94576f715fa7cd7135b00ab0e0cf032



Embedding for query generated successfully
Similarity search response: []

**Reply by razataiab** [reply]
I've implemented collection loading

**Reply by shanghaikid** [reply]
It looks like your data isn't inserted successfully. can you print the data to insert?

**Reply by razataiab** [reply]
Milvus.js:

const { MilvusClient, DataType } = require('@zilliz/milvus2-sdk-node');

const client = new MilvusClient({ address: 'http://localhost:19530' });

function generateUniqueCollectionName() {
    const timestamp = Date.now();
    return `embedding_${timestamp}`;
}

async function storeEmbeddings(collectionName, filename, embeddings) {
    try {
        console.log(`Embedding length: ${embeddings[0].length}`);

        await client.createCollection({
            collection_name: collectionName,
            fields: [
                { name: "my_id", data_type: DataType.Int64, is_primary_key: true, auto_id: true },
                { name: "filename", data_type: DataType.VarChar, max_length: 256 },
                { name: "embedding", data_type: DataType.FloatVector, dim: embeddings[0].length }
            ],
        });
        console.log(`Collection ${collectionName} created`);

        // Prepare the row-based data for insertion
        const rowData = embeddings.map(embedding => ({
            filename: filename,
            embedding: embedding
        }));

        const insert = await client.insert({
            collection_name: collectionName,
            data: rowData
        });
    
    console.log('insert result: ', insert);

        await client.createIndex({
            collection_name: collectionName,
            field_name: "embedding",
            index_type: "IVF_FLAT",
            metric_type: "L2",
            index_name: "vector_index",
            params: { "nlist": 128 }
        });

        console.log("Embeddings stored successfully in Milvus");
    } catch (error) {
        console.error('Failed to store embeddings in Milvus:', error);
    }
}

module.exports = {
    storeEmbeddings,
    generateUniqueCollectionName,
};



Terminal Output:


Fields data prepared for insertion: {
  filename: [
    '934d051e05e71b10457686d74ad749db',
    '934d051e05e71b10457686d74ad749db',
    '934d051e05e71b10457686d74ad749db',
    '934d051e05e71b10457686d74ad749db',
    '934d051e05e71b10457686d74ad749db',
    '934d051e05e71b10457686d74ad749db',
    '934d051e05e71b10457686d74ad749db',
    '934d051e05e71b10457686d74ad749db',
    '934d051e05e71b10457686d74ad749db',
    '934d051e05e71b10457686d74ad749db'
  ],
  embedding: [
    [
         0.02498547, -0.013231517,      0.01211336,    0.042436697,
      ... 1436 more items
    ],
    [
       -0.025081575,  -0.027821356, -0.0013277021, -0.013427577,  0.036186293,
      ... 1436 more items
    ],
    [
       -0.008697558,  0.0014513106,   0.029926196,    0.04336413,   -0.03962679,
      ... 1436 more items
    ],
    [
          0.03758286,   0.031041719,   0.031701572,    0.066272065,
      ... 1436 more items
    ],
    [
       -0.007097355,  0.037363816,    0.022891024,   0.004752637,
      ... 1436 more items
    ],
    [
        -0.023545034,  0.033187028,   0.0067432486,   0.023927437,
      ... 1436 more items
    ],
    [
        0.012714208, -0.00021340512,   0.007527465,    0.05611788,
      ... 1436 more items
    ],
    [
         0.03297819, -0.0030862135,   0.007556281,    0.034237567,
      ... 1436 more items
    ],
    [
          0.0335562, 0.0035057182,   0.020067923,   0.05603666,   0.044494864,
      ... 1436 more items
    ],
    [
         0.02197845,   0.046539366,  0.024313658,    0.03203359,   0.026470294,
      ... 1436 more items
    ]
  ]
}
Failed to store embeddings in Milvus: Error: The type of the `data or field_data` should be an array and length > 0.

**Reply by shanghaikid** [reply]
Nodejs doesn't support insert by column. so please build row based data. like this.

```javascript
[
{filename: 'some-file-name1', embedding: [...]}
{filename: 'some-file-name2', embedding: [...]}
...
]
```

**Reply by razataiab** [reply]
I tried, still the same error is recurring

insert result:  {
  succ_index: [],
  err_index: [
    0, 1, 2, 3, 4,
    5, 6, 7, 8, 9
  ],
  status: {
    extra_info: {},
    error_code: 'UnexpectedError',
    reason: 'the num_rows(0) of 0th field is not equal to passed NumRows(10)',
    code: 0,
    retriable: false,
    detail: ''
  },
  IDs: null,
  acknowledged: false,
  insert_cnt: '0',
  delete_cnt: '0',
  upsert_cnt: '0',
  timestamp: '0'
}






const { MilvusClient, DataType } = require('@zilliz/milvus2-sdk-node');

const client = new MilvusClient({ address: 'http://localhost:19530' });

function generateUniqueCollectionName() {
    const timestamp = Date.now();
    return `embedding_${timestamp}`;
}

async function storeEmbeddings(collectionName, filename, embeddings) {
    try {
        console.log(`Embedding length: ${embeddings[0].length}`);

        await client.createCollection({
            collection_name: collectionName,
            fields: [
                { name: "my_id", data_type: DataType.Int64, is_primary_key: true, auto_id: true },
                { name: "filename", data_type: DataType.VarChar, max_length: 256 },
                { name: "embedding", data_type: DataType.FloatVector, dim: embeddings[0].length }
            ],
        });
        console.log(`Collection ${collectionName} created`);

        // Prepare the row-based data for insertion
        const rowData = embeddings.map(embedding => ({
            filename: filename,
            embedding: embedding
        }));

        const insert = await client.insert({
            collection_name: collectionName,
            data: rowData
        });
    
    console.log('insert result: ', insert);

        await client.createIndex({
            collection_name: collectionName,
            field_name: "embedding",
            index_type: "IVF_FLAT",
            metric_type: "L2",
            index_name: "vector_index",
            params: { "nlist": 128 }
        });

        console.log("Embeddings stored successfully in Milvus");
    } catch (error) {
        console.error('Failed to store embeddings in Milvus:', error);
    }
}

module.exports = {
    storeEmbeddings,
    generateUniqueCollectionName,
};

**Reply by shanghaikid** [reply]
https://github.com/milvus-io/milvus-sdk-node/blob/main/examples/milvus/HelloMilvus.ts 

this is working example, replace the address with yours, then run `ts-node` HelloMilvus， it should print something like this. 

![image](https://github.com/milvus-io/milvus/assets/185051/a845e169-d28f-4baa-bd56-17d229695dee)

**Reply by razataiab** [reply]
razataiab@Razas-MacBook-Pro utils % ts-node hellomilvus      
Node client is initialized.
Create collection is finished. {
  extra_info: {},
  error_code: 'Success',
  reason: '',
  code: 0,
  retriable: false,
  detail: ''
}
Data is inserted.
Index is created {
  extra_info: {},
  error_code: 'UnexpectedError',
  reason: 'invalid index params: [key:"metric_type" value:"L2" ]',
  code: 0,
  retriable: false,
  detail: ''
}
TypeError: Cannot read properties of undefined (reading 'error_code')
    at MilvusClient.<anonymous> (/Users/razataiab/Desktop/custom-rag/node_modules/@zilliz/milvus2-sdk-node/milvus/grpc/Collection.ts:469:22)
    at Generator.next (<anonymous>)
    at fulfilled (/Users/razataiab/Desktop/custom-rag/node_modules/@zilliz/milvus2-sdk-node/dist/milvus/grpc/Collection.js:5:58)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    
    
    
    
    
    almost there, all good uptill load collection

**Reply by shanghaikid** [reply]
what is your milvus version?

**Comment by razataiab** [comment]
Hey, the data is now being added to the collection in milvus, ive set the
dim to 10, but the similarity vector result is still null

**Reply by shanghaikid** [reply]
show me your code

**Reply by shanghaikid** [reply]
do you use `client.loadCollectionSync` ?

**Comment by razataiab** [comment]
I use the load collection function.

const { MilvusClient , ConsistencyLevelEnum } = require('
@zilliz/milvus2-sdk-node');
const { getCollectionName } = require('../utils/config'); // Import the
config module

const client = new MilvusClient({ address: 'localhost:19530' });

async function similaritySearch(queryEmbedding) {
try {
const collectionName = getCollectionName(); // Retrieve the collection name
if (!collectionName) {
throw new Error('Collection name is not set.');
}

await client.loadCollection({
collection_name: collectionName
});

const searchParams = {
anns_field: "vector",
topk: "3",
metric_type: "L2",
};

// Conduct similarity search
const results = await client.search({
collection_name: collectionName,
vectors: [queryEmbedding],
search_params: searchParams,
vector_type: 101,
});

// Check if response is defined and has results
if (!results || !results.results) {
throw new Error('Search response is undefined or does not contain results');
}

console.log('Similarity search response:', results.results);
return results.results;
} catch (error) {
console.error('Error during similarity search:', error);
throw error;
}
}

module.exports = { similaritySearch };

**Reply by shanghaikid** [reply]
please replace `loadCollection` with `loadCollectionSync`

**Reply by shanghaikid** [reply]
in the next version of node sdk, I will make `loadCollection` equal `loadCollectionSync`

**Reply by razataiab** [reply]
Deadline exceeded after 15.001s,metadata filters: 0.001s,remote_addr=[::1]:19530


it's slow I guess, I've made the load collection change

**Reply by razataiab** [reply]
{
  code: 4,
  details: 'Deadline exceeded after 15.001s,name resolution: 0.003s,LB pick: 0.001s,remote_addr=[::1]:19530',
  metadata: Metadata { internalRepr: Map(0) {}, options: {} }
}


the error is repetitive

**Reply by shanghaikid** [reply]
it looks like your server is not connected. can you check if your server is working?

Regarding the node code, you can follow this doc: 
https://milvus.io/docs/single-vector-search.md

**Comment by phamtrung0633** [comment]
Hi, I know why, it is not something to do with Milvus itself but to do with the local Milvus db of you. I encounter the same problem, I don't know how to solve to be honest for local Milvus, but i used 'zilliz.com' for my milvus database and it is all good. Maybe consider using this cloud solution until somebody found out the reason

**Reply by shanghaikid** [reply]
What's your problem specifically? I don't think here there would be any difference in terms of local milvus or Zilliz Cloud.

**Reply by phamtrung0633** [reply]
My problem is the exact same one that happened to him, changing from localaddress to zilliz cloud address solves instantly

**Reply by shanghaikid** [reply]
Ok, please upgrade the node sdk. now `loadCollection` is equal to `loadCollectionSync`, it should solve your problem as well.  Zilliz Cloud loads collection for you automatically.

**Reply by phamtrung0633** [reply]
I actually did and I used loadcollectionsync, it is still the same issue, Zilliz solves it instantly ..
